{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    \"C:\\\\Users\\\\Wyss User\\\\AppData\\\\Local\\\\Packages\\\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\\\LocalCache\\\\local-packages\\\\Python311\\\\site-packages\"\n",
    ")\n",
    "\n",
    "import biolib\n",
    "import gzip\n",
    "import math\n",
    "import os\n",
    "import requests\n",
    "import pathlib\n",
    "import tspex\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "from io import StringIO\n",
    "from IPython.display import Image\n",
    "from matplotlib_venn import venn2\n",
    "from matplotlib_venn import venn3\n",
    "from scipy import stats\n",
    "from scipy.stats import f_oneway\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ALL_FASTA_OUTPUT_PATH = \"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\\\\all_fasta.fasta\"\n",
    "# ALL_TARGETS_OUTPUT_DIRECTORY = (\n",
    "#     \"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\\\\tmhmm_all_targets\"\n",
    "# )\n",
    "\n",
    "ASSAY_LIST_PATH = \"assay-list-olink-explore-3072.xlsx\"\n",
    "BRAIN_RNA_SEQ_PATH = \"brain_rna_seq.csv\"\n",
    "HGNC_IDS = (\n",
    "    \"https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt\"\n",
    ")\n",
    "\n",
    "PLATE_LAYOUT_PATH = \"Plate Layout.xlsx\"\n",
    "RAW_DATA_FILE = \"LCSET_28343_10-4-2023_EXTENDED_NPX_2023-10-20.csv\"\n",
    "# TMHMM_OUTPUT = (\n",
    "#     \"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\\\\tmhmm_all_targets\\\\TMRs.gff3\"\n",
    "# )\n",
    "UNIPROT_TO_FASTA_PATH = \"uniprotkb_proteome_UP000005640_2023_11_20.fasta.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_list = pd.read_excel(ASSAY_LIST_PATH)\n",
    "\n",
    "\n",
    "# Map UniProt IDs to the corresponding FASTA sequences\n",
    "def parse_gz_file(file_path):\n",
    "    protein_dict = {}\n",
    "    current_uniprot_id = None\n",
    "    current_sequence = \"\"\n",
    "    with gzip.open(file_path, \"rt\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if current_uniprot_id is not None:\n",
    "                    protein_dict[current_uniprot_id] = current_sequence\n",
    "                    current_sequence = \"\"\n",
    "                if \"|\" in line:\n",
    "                    current_uniprot_id = line.split(\"|\")[1].strip()\n",
    "                else:\n",
    "                    print(f\"Skipping line without expected format: {line}\")\n",
    "                    current_uniprot_id = None\n",
    "            else:\n",
    "                current_sequence += line\n",
    "        if current_uniprot_id is not None:\n",
    "            protein_dict[current_uniprot_id] = current_sequence\n",
    "    return protein_dict\n",
    "\n",
    "protein_data = parse_gz_file(UNIPROT_TO_FASTA_PATH)\n",
    "\n",
    "assay_list[\"Sequence\"] = assay_list[\"UniProt ID\"].map(\n",
    "    lambda x: protein_data.get(x, \"N/A\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create a FASTA file containing the information for the OLINK targets\n",
    "\n",
    "# def export_fasta(output_path, df):\n",
    "#     with open(output_path, \"w\") as output_file:\n",
    "#         for index, row in df.iterrows():\n",
    "#             output_file.write(f\">{row[\"UniProt ID\"]}\\n\")\n",
    "#             output_file.write(f\"{row[\"Sequence\"]}\\n\")\n",
    "# export_fasta(ALL_FASTA_OUTPUT_PATH, assay_list)\n",
    "\n",
    "# # Use the TMHMM deep learning model to predict the localication of each of the OLINK targets\n",
    "# deeptmhmm = biolib.load(\"DTU/DeepTMHMM\")\n",
    "# def tmhmm_localization(targets, output_directory):\n",
    "#     if not os.path.exists(output_directory):\n",
    "#         os.makedirs(output_directory)\n",
    "#     os.chdir(output_directory)\n",
    "#     with open(\"query.fasta\", \"w\") as fasta_file:\n",
    "#         for _, row in targets.iterrows():\n",
    "#             sequence = row[\"Sequence\"]\n",
    "#             uniprot_id = row[\"UniProt ID\"]\n",
    "#             fasta_line = f\">{uniprot_id}\\n{sequence}\\n\"\n",
    "#             fasta_file.write(fasta_line)\n",
    "#         biolib.utils.STREAM_STDOUT = True\n",
    "#         deeptmhmm_job = deeptmhmm.cli(\n",
    "#             args=\"--fasta query.fasta\", machine=\"local\"\n",
    "#         )\n",
    "#         deeptmhmm_job.save_files(output_directory)\n",
    "\n",
    "# tmhmm_localization(assay_list, ALL_TARGETS_OUTPUT_DIRECTORY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create lists to store data\n",
    "uniprot_ids = []\n",
    "region_locations = []\n",
    "region_starts = []\n",
    "region_ends = []\n",
    "\n",
    "# Import the TMHMM output\n",
    "with open(TMHMM_OUTPUT, \"r\") as gff_file:\n",
    "    for line in gff_file: \n",
    "        if not line.startswith(\"//\"):\n",
    "            columns = line.strip().split(\"\\t\")\n",
    "            if len(columns) >= 4:\n",
    "                uniprot_id = columns[0]\n",
    "                region_location = columns[1]\n",
    "                region_start = int(columns[2])\n",
    "                region_end = int(columns[3])\n",
    "\n",
    "                uniprot_ids.append(uniprot_id)\n",
    "                region_locations.append(region_location)\n",
    "                region_starts.append(region_start)\n",
    "                region_ends.append(region_end)\n",
    "\n",
    "# Create a new dataframe\n",
    "data = {\n",
    "    \"UniProt ID\": uniprot_ids,\n",
    "    \"Region Location\": region_locations,\n",
    "    \"Region Start\": region_starts,\n",
    "    \"Region End\": region_ends,\n",
    "}\n",
    "\n",
    "tmhmm_df = pd.DataFrame(data)\n",
    "\n",
    "tm_tmhmm = tmhmm_df[tmhmm_df[\"Region Location\"] == \"TMhelix\"]\n",
    "outside_tmhmm = tmhmm_df[tmhmm_df[\"Region Location\"] == \"outside\"]\n",
    "inside_tmhmm = tmhmm_df[tmhmm_df[\"Region Location\"] == \"inside\"]\n",
    "\n",
    "tm_uniprots = set(tm_tmhmm[\"UniProt ID\"])\n",
    "inside_uniprots = set(inside_tmhmm[\"UniProt ID\"])\n",
    "outside_uniprots = set(outside_tmhmm[\"UniProt ID\"])\n",
    "\n",
    "inside_uniprots = inside_uniprots - tm_uniprots - outside_uniprots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def export_fasta(output_path, df):\n",
    "#     with open(output_path, \"w\") as output_file:\n",
    "#         for index, row in df.iterrows():\n",
    "#             output_file.write(f\">{row[\"UniProt ID\"]}\\n\")\n",
    "#             output_file.write(f\"{row[\"Sequence\"]}\\n\")\n",
    "\n",
    "def tmhmm_localization(assay_list, output_directory):\n",
    "    deeptmhmm = biolib.load(\"DTU/DeepTMHMM\")\n",
    "    with open(\"query.fasta\", \"w\") as fasta_file:\n",
    "        for _, row in assay_list.iterrows():\n",
    "            fasta_line = f\">{assay_list['UniProt ID']}\\n{assay_list['Sequence']}\\n\"\n",
    "            fasta_file.write(fasta_line)\n",
    "        deeptmhmm_job = deeptmhmm.cli(\n",
    "            args=\"--fasta query.fasta\", machine=\"local\"\n",
    "        )\n",
    "        deeptmhmm_job.save_files(output_directory)\n",
    "\n",
    "\n",
    "def identify_localization(assay_list, region, output_directory = \"olink_output\"):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    assay_list : pandas.DataFrame\n",
    "        DataFrame with columns called 'UniProt ID' and 'Sequence'\n",
    "    region : {'TMhelix', 'inside', 'outside', 'internal'}\n",
    "        Subcellular region requested. Options:\n",
    "          - 'TMhelix': transmembrane proteins\n",
    "          - 'inside': at least some of the protein is inside the cell/EV\n",
    "          - 'outside': at least some of the protein is outside the cell/EV\n",
    "          - 'internal': the protein is only internal, no transmembrane or outside\n",
    "    \"\"\"\n",
    "    output_directory_path = pathlib.Path(output_directory)\n",
    "    if not os.path.exists(output_directory_path):\n",
    "        os.makedirs(output_directory_path)\n",
    "    # os.chdir(output_directory)\n",
    "    if not os.path.exists(output_directory_path / \"TMRs.gff3\"):\n",
    "        tmhmm_localization(assay_list, output_directory_path)\n",
    "    localization_df = pd.read_csv(output_directory_path / \"TMRs.gff3\", sep = \"\\t\", comment = \"#\", names = [\"uniprot_id\", \"region_location\", \"region_start\", \"region_end\", 0, 1, 2, 3])\n",
    "    localization_df = localization_df[localization_df[\"uniprot_id\"] != \"//\"].dropna(axis = 1)\n",
    "    get_regional_uniprots = lambda region: set(localization_df[localization_df[\"region_location\"] == region][\"uniprot_id\"])\n",
    "    if region == \"internal\":\n",
    "        tm_uniprots = get_regional_uniprots(\"TMhelix\")\n",
    "        outside_uniprots = get_regional_uniprots(\"outside\")\n",
    "        inside_uniprots = get_regional_uniprots(\"inside\")\n",
    "        return (inside_uniprots - tm_uniprots - outside_uniprots)\n",
    "    else: \n",
    "         return get_regional_uniprots(region)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data to plot\n",
    "labels = [\"Transmembrane\", \"Inside\", \"Outside\"]\n",
    "sizes = [774, 1300, 823]\n",
    "colors = ['purple', 'lightcoral', 'lightskyblue']\n",
    "\n",
    "# Plot\n",
    "plt.pie(sizes, labels=labels, colors=colors, autopct='%1.1f%%')\n",
    "plt.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.\n",
    "plt.title(\"OLINK Target Localization\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define necessary functions and import files\n",
    "def calculate_mean(df):\n",
    "    return df.assign(Mean=df.mean(axis=1, numeric_only=True))\n",
    "\n",
    "brain_rna_seq = pd.read_csv(BRAIN_RNA_SEQ_PATH)\n",
    "\n",
    "# Map the HGNC IDs to the UniProt IDs\n",
    "# This database that maps HGNC IDs to UniProt IDs doesn\"t have all the UniProt IDs we need, resulting in some targets being dropped unnecessarily\n",
    "response = requests.get(HGNC_IDS)\n",
    "content = response.text\n",
    "file_like_object = StringIO(content)\n",
    "uniprot_ids = pd.read_csv(file_like_object, sep=\"\\t\", low_memory=False)\n",
    "\n",
    "brain_rna_seq = pd.merge(brain_rna_seq, uniprot_ids, left_on = \"id\", right_on = \"hgnc_id\", how = \"inner\")\n",
    "\n",
    "# Find the mean expression of each cell type \n",
    "mature_astrocytes_cols = [\n",
    "    \"gene_id\",\n",
    "    \"id\",\n",
    "    \"uniprot_ids\",\n",
    "    \"astrocytes_mature_1\",\n",
    "    \"astrocytes_mature_2\",\n",
    "    \"astrocytes_mature_3\",\n",
    "    \"astrocytes_mature_4\",\n",
    "    \"astrocytes_mature_5\",\n",
    "    \"astrocytes_mature_6\",\n",
    "    \"astrocytes_mature_7\",\n",
    "    \"astrocytes_mature_8\",\n",
    "    \"astrocytes_mature_9\",\n",
    "    \"astrocytes_mature_10\",\n",
    "    \"astrocytes_mature_11\",\n",
    "    \"astrocytes_mature_12\",\n",
    "]\n",
    "mature_astrocytes = brain_rna_seq[mature_astrocytes_cols]\n",
    "\n",
    "endothelial_cols = [\"gene_id\", \"id\", \"uniprot_ids\", \"endothelial_1\", \"endothelial_2\"]\n",
    "endothelial = brain_rna_seq[endothelial_cols]\n",
    "\n",
    "microglia_cols = [\"gene_id\", \"id\", \"uniprot_ids\", \"microglla_1\", \"microglla_2\", \"microglla_3\"]\n",
    "microglia = brain_rna_seq[microglia_cols]\n",
    "\n",
    "oligodendrocyte_cols = [\n",
    "    \"gene_id\",\n",
    "    \"id\",\n",
    "    \"uniprot_ids\",\n",
    "    \"oligodendrocytes_1\",\n",
    "    \"oligodendrocytes_2\",\n",
    "    \"oligodendrocytes_3\",\n",
    "    \"oligodendrocytes_4\",\n",
    "    \"oligodendrocytes_5\",\n",
    "]\n",
    "oligodendrocytes = brain_rna_seq[oligodendrocyte_cols]\n",
    "\n",
    "neurons_cols = [\"gene_id\", \"id\", \"uniprot_ids\", \"neurons_1\"]\n",
    "neurons = brain_rna_seq[neurons_cols]\n",
    "\n",
    "neurons = calculate_mean(neurons)\n",
    "oligodendrocytes = calculate_mean(oligodendrocytes)\n",
    "microglia = calculate_mean(microglia)\n",
    "endothelial = calculate_mean(endothelial)\n",
    "mature_astrocytes = calculate_mean(mature_astrocytes)\n",
    "\n",
    "neurons = neurons[[\"id\", \"uniprot_ids\", \"Mean\"]]\n",
    "oligodendrocytes = oligodendrocytes[[\"id\", \"uniprot_ids\", \"Mean\"]]\n",
    "microglia = microglia[[\"id\", \"uniprot_ids\", \"Mean\"]]\n",
    "endothelial = endothelial[[\"id\", \"uniprot_ids\", \"Mean\"]]\n",
    "mature_astrocytes = mature_astrocytes[[\"id\", \"uniprot_ids\", \"Mean\"]]\n",
    "\n",
    "neurons.rename(columns = {\"id\": \"hgnc_id\", \"uniprot_ids\": \"uniprot_id\", \"Mean\": \"Neurons\"}, inplace = True)\n",
    "oligodendrocytes.rename(columns = {\"id\": \"hgnc_id\", \"uniprot_ids\": \"uniprot_id\", \"Mean\": \"Oligodendrocytes\"}, inplace = True)\n",
    "microglia.rename(columns = {\"id\": \"hgnc_id\", \"uniprot_ids\": \"uniprot_id\", \"Mean\": \"Microglia\"}, inplace = True)\n",
    "endothelial.rename(columns = {\"id\": \"hgnc_id\", \"uniprot_ids\": \"uniprot_id\", \"Mean\": \"Endothelial\"}, inplace = True)\n",
    "mature_astrocytes.rename(columns = {\"id\": \"hgnc_id\", \"uniprot_ids\": \"uniprot_id\", \"Mean\": \"Astrocytes\"}, inplace = True)\n",
    "\n",
    "all_cells = pd.merge(neurons, oligodendrocytes, on = \"hgnc_id\")\n",
    "all_cells = all_cells[[\"hgnc_id\", \"uniprot_id_x\", \"Neurons\", \"Oligodendrocytes\"]]\n",
    "all_cells = pd.merge(all_cells, microglia, on = \"hgnc_id\")\n",
    "all_cells = all_cells[[\"hgnc_id\", \"uniprot_id_x\", \"Neurons\", \"Oligodendrocytes\", \"Microglia\"]]\n",
    "all_cells = pd.merge(all_cells, endothelial, on = \"hgnc_id\")\n",
    "all_cells = all_cells[[\"hgnc_id\", \"uniprot_id_x\", \"Neurons\", \"Oligodendrocytes\", \"Microglia\", \"Endothelial\"]]\n",
    "all_cells = pd.merge(all_cells, mature_astrocytes, on = \"hgnc_id\")\n",
    "all_cells = all_cells[[\"hgnc_id\", \"uniprot_id_x\", \"Neurons\", \"Oligodendrocytes\", \"Microglia\", \"Endothelial\", \"Astrocytes\"]]\n",
    "\n",
    "all_cells.drop_duplicates(subset=\"uniprot_id_x\", inplace=True)\n",
    "all_cells.dropna(subset=[\"uniprot_id_x\"], inplace=True)\n",
    "all_cells.set_index(\"uniprot_id_x\", inplace = True)\n",
    "\n",
    "# Determine the tissue specificity of each gene\n",
    "expression_data = all_cells.loc[:, [\"Astrocytes\", \"Endothelial\", \"Microglia\", \"Neurons\", \"Oligodendrocytes\"]]\n",
    "expression_data = expression_data.loc[(expression_data > 0).any(axis=1)]\n",
    "tso_gen = tspex.TissueSpecificity(expression_data, 'tau', log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_df = pd.DataFrame(tso_gen.tissue_specificity, columns = [\"Tau Score\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.xlabel(\"Tau Score\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.title(\"Distribution of Tau Values\")\n",
    "\n",
    "plt.hist(list(tso_df[\"Tau Score\"]), bins = 20, color = \"black\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_df_filtered = tso_df[tso_df[\"Tau Score\"] > 0.85]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_filtered = list(tso_df_filtered.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data_filtered = expression_data.loc[expression_data.index.isin(tso_filtered)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_ind = tspex.TissueSpecificity(expression_data_filtered, 'zscore', log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_ind.plot_heatmap(threshold=0.8, sort_genes=True, use_zscore=True, gene_names=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(type(tso_ind))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_ind_df = pd.DataFrame(tso_ind.tissue_specificity)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tso_ind_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "astrocyte_uniprot_ids = tso_ind_df[tso_ind_df[\"Astrocytes\"] > 0.99].index.tolist()\n",
    "microglia_uniprot_ids = tso_ind_df[tso_ind_df[\"Microglia\"] > 0.99].index.tolist()\n",
    "neuron_uniprot_ids = tso_ind_df[tso_ind_df[\"Neurons\"] > 0.99].index.tolist()\n",
    "oligodendrocyte_uniprot_ids = tso_ind_df[tso_ind_df[\"Oligodendrocytes\"] > 0.99].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "venn3(\n",
    "    [set(astrocyte_uniprot_ids), set(microglia_uniprot_ids), set(neuron_uniprot_ids)],\n",
    "    set_labels = (\"Astrocytes\", \"Microglia\", \"Neurons\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "venn3(\n",
    "    [set(oligodendrocyte_uniprot_ids), set(microglia_uniprot_ids), set(neuron_uniprot_ids)],\n",
    "    set_labels = (\"Oligodendrocytes\", \"Microglia\", \"Neurons\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "venn2(\n",
    "    [set(oligodendrocyte_uniprot_ids), set(astrocyte_uniprot_ids)],\n",
    "    set_labels = (\"Oligodendrocytes\", \"Astrocytes\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_dict = {\n",
    "    len(astrocyte_uniprot_ids): \"Astrocytes\",\n",
    "    len(microglia_uniprot_ids): \"Microglia\",\n",
    "    len(neuron_uniprot_ids): \"Neurons\",\n",
    "    len(oligodendrocyte_uniprot_ids): \"Oligodendrocytes\"\n",
    "}\n",
    "\n",
    "x_labels = list(expression_dict.values())\n",
    "y_labels = list(expression_dict.keys())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "plt.xlabel(\"Cell Type\")\n",
    "plt.ylabel(\"Number of Genes\")\n",
    "plt.title(\"Cell Type Specificity: Z-Score\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = [\"Microglia\", \"Astrocytes\", \"Oligodendrocytes\", \"Neurons\"]\n",
    "sizes = [635, 975, 295, 843]\n",
    "colors = ['orchid', 'lightcoral', 'lightskyblue', \"slateblue\"]\n",
    "\n",
    "plt.pie(sizes, labels=labels, colors=colors, autopct='%1.1f%%')\n",
    "plt.axis('equal') \n",
    "plt.title(\"Cell-Type Specificity: Z-Score\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIMPORTANT_COLUMNS = [\n",
    "    \"Index\",\n",
    "    \"OlinkID\",\n",
    "    \"Panel_Lot_Nr\",\n",
    "    \"PlateID\",\n",
    "    \"WellID\",\n",
    "    \"Normalization\",\n",
    "    \"Processing_StartDate\",\n",
    "    \"Processing_EndDate\",\n",
    "    \"AnalyzerID\",\n",
    "    \"ExploreVersion\",\n",
    "]\n",
    "\n",
    "PANELS = [\n",
    "    \"Cardiometabolic\",\n",
    "    \"Cardiometabolic_II\",\n",
    "    \"Inflammation\",\n",
    "    \"Inflammation_II\",\n",
    "    \"Neurology\",\n",
    "    \"Neurology_II\",\n",
    "    \"Oncology\",\n",
    "    \"Oncology_II\",\n",
    "]\n",
    "\n",
    "# Clean up the raw OLINK data file and map the sample locations to their barcodes\n",
    "plate_layout = pd.read_excel(PLATE_LAYOUT_PATH, index_col=0)\n",
    "raw_data = pd.read_csv(RAW_DATA_FILE, sep=\";\")\n",
    "raw_data = raw_data.drop(UNIMPORTANT_COLUMNS, axis=1)\n",
    "\n",
    "data = pd.merge(raw_data, plate_layout, how=\"left\", on=\"SampleID\")\n",
    "\n",
    "# Normalize the data according to plate negative controls\n",
    "ctrl_dict = {}\n",
    "for panel in PANELS:\n",
    "    df = data[(data[\"Panel\"] == panel)]\n",
    "    plate_ctrl = df[\n",
    "        (df[\"Sample_Type\"] == \"PLATE_CONTROL\")\n",
    "        & (df[\"Assay\"] == \"Amplification control 1\")\n",
    "    ][\"NPX\"].median()\n",
    "    ctrl_dict[panel] = plate_ctrl\n",
    "data[\"Delta\"] = data.apply(\n",
    "    lambda row: row[\"NPX\"] - ctrl_dict[row[\"Panel\"]],\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "# Linearize the NPX values\n",
    "data[\"Linear Delta\"] = data[\"Delta\"].map(lambda x: 2**x)\n",
    "\n",
    "# Identify and drop replicate assays\n",
    "vc = data[[\"SampleID\", \"Sample_Type\", \"UniProt\"]].value_counts()\n",
    "vc = vc[vc > 1]\n",
    "\n",
    "unique_uniprot_values = vc.index.get_level_values(\"UniProt\").unique()\n",
    "uniprot_list = unique_uniprot_values.tolist()\n",
    "\n",
    "raw_data_no_reps = data[~data[\"UniProt\"].isin(uniprot_list)].copy()\n",
    "\n",
    "# Create a tidy dataframe for the delta values\n",
    "tidy_data = raw_data_no_reps[raw_data_no_reps[\"Sample_Type\"] == \"SAMPLE\"].pivot(\n",
    "    columns=\"UniProt\", index=[\"SampleID\", \"Health\", \"Sample\"], values=\"Linear Delta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mann-Whitney U Test to determine significant difference between AD/MCI/Healthy samples in all fractions\n",
    "\n",
    "assay_list = list(tidy_data.columns)\n",
    "\n",
    "healthy_df = tidy_data[tidy_data.index.get_level_values(\"Health\") == \"Healthy\"]\n",
    "mci_df = tidy_data[tidy_data.index.get_level_values(\"Health\") == \"MCI\"]\n",
    "ad_df = tidy_data[tidy_data.index.get_level_values(\"Health\") == \"AD\"]\n",
    "\n",
    "ad_healthy_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "ad_mci_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "mci_healthy_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "\n",
    "for assay in assay_list:\n",
    "    healthy_assays = list(healthy_df[assay])\n",
    "    mci_assays = list(mci_df[assay])\n",
    "    ad_assays = list(ad_df[assay])\n",
    "\n",
    "    ad_healthy_statistic, ad_healthy_p_value = mannwhitneyu(ad_assays, healthy_assays)\n",
    "    ad_mci_statistic, ad_mci_p_value = mannwhitneyu(ad_assays, mci_assays)\n",
    "    mci_healthy_statistic, mci_healthy_p_value = mannwhitneyu(\n",
    "        mci_assays, healthy_assays\n",
    "    )\n",
    "\n",
    "    ad_healthy_result_row = pd.DataFrame(\n",
    "        {\n",
    "            \"Assay\": [assay],\n",
    "            \"p-value\": [ad_healthy_p_value],\n",
    "            \"Statistic\": [ad_healthy_statistic],\n",
    "        }\n",
    "    )\n",
    "    ad_mci_result_row = pd.DataFrame(\n",
    "        {\"Assay\": [assay], \"p-value\": [ad_mci_p_value], \"Statistic\": [ad_mci_statistic]}\n",
    "    )\n",
    "    mci_healthy_result_row = pd.DataFrame(\n",
    "        {\n",
    "            \"Assay\": [assay],\n",
    "            \"p-value\": [mci_healthy_p_value],\n",
    "            \"Statistic\": [mci_healthy_statistic],\n",
    "        }\n",
    "    )\n",
    "\n",
    "    ad_healthy_mannwhitneyu = pd.concat(\n",
    "        [ad_healthy_mannwhitneyu, ad_healthy_result_row], ignore_index=True\n",
    "    )\n",
    "    ad_mci_mannwhitneyu = pd.concat(\n",
    "        [ad_mci_mannwhitneyu, ad_mci_result_row], ignore_index=True\n",
    "    )\n",
    "    mci_healthy_mannwhitneyu = pd.concat(\n",
    "        [mci_healthy_mannwhitneyu, mci_healthy_result_row], ignore_index=True\n",
    "    )\n",
    "\n",
    "ad_healthy_sig_dif = ad_healthy_mannwhitneyu[ad_healthy_mannwhitneyu[\"p-value\"] < 0.05]\n",
    "ad_mci_sig_dif = ad_mci_mannwhitneyu[ad_mci_mannwhitneyu[\"p-value\"] < 0.05]\n",
    "mci_healthy_sig_dif = mci_healthy_mannwhitneyu[\n",
    "    mci_healthy_mannwhitneyu[\"p-value\"] < 0.05\n",
    "]\n",
    "\n",
    "ad_healthy_sig_dif_assays = set(ad_healthy_sig_dif[\"Assay\"])\n",
    "ad_mci_sig_dif_assays = set(ad_mci_sig_dif[\"Assay\"])\n",
    "mci_healthy_sig_dif_assays = set(mci_healthy_sig_dif[\"Assay\"])\n",
    "\n",
    "venn3(\n",
    "    [ad_healthy_sig_dif_assays, ad_mci_sig_dif_assays, mci_healthy_sig_dif_assays],\n",
    "    set_labels=(\"AD v. Healthy\", \"AD v. MCI\", \"MCI v. Healthy\"),\n",
    ")\n",
    "plt.title(\"Assays with Significant Differences Between Samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mann-Whitney U Test to determine significant difference between AD/MCI/Healthy samples in EV/secreted fractions\n",
    "\n",
    "assay_list = list(tidy_data.columns)\n",
    "\n",
    "ad_healthy_ev_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "ad_mci_ev_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "mci_healthy_ev_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "\n",
    "ad_healthy_secreted_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "ad_mci_secreted_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "mci_healthy_secreted_mannwhitneyu = pd.DataFrame(columns=[\"Assay\", \"p-value\", \"Statistic\"])\n",
    "\n",
    "for assay in assay_list:\n",
    "    tidy_data_assay = tidy_data[assay]\n",
    "    if not tidy_data_assay.isna().all():\n",
    "        healthy_assay_ev_peak = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"9\"))\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"10\"))\n",
    "            ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "        mci_assay_ev_peak = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"9\"))\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"10\"))\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "        mci_assay_ev_peak = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"9\"))\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"10\"))\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        ad_assay_ev_peak = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"9\"))\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\").str.contains(\"10\"))\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"AD\")\n",
    "        healthy_assay_ev_values = tidy_data_assay.loc[healthy_assay_ev_peak].values\n",
    "        mci_assay_ev_values = tidy_data_assay.loc[mci_assay_ev_peak].values\n",
    "        ad_assay_ev_values = tidy_data_assay.loc[ad_assay_ev_peak].values\n",
    "        if not (np.all(np.isnan(healthy_assay_ev_peak))) or not (np.all(np.isnan(mci_assay_ev_peak))) or not (np.all(np.isnan(ad_assay_ev_peak))):\n",
    "            ad_healthy_ev_statistic, ad_healthy_ev_p_value = mannwhitneyu(\n",
    "                ad_assay_ev_values, healthy_assay_ev_values\n",
    "            )\n",
    "            ad_mci_ev_statistic, ad_mci_ev_p_value = mannwhitneyu(\n",
    "                ad_assay_ev_values, mci_assay_ev_values\n",
    "            )\n",
    "            mci_healthy_ev_statistic, mci_healthy_ev_p_value = mannwhitneyu(\n",
    "                mci_assay_ev_values, healthy_assay_ev_values\n",
    "            )\n",
    "            ad_healthy_ev_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [ad_healthy_ev_p_value],\n",
    "                    \"Statistic\": [ad_healthy_ev_statistic],\n",
    "                }\n",
    "            )\n",
    "            ad_mci_ev_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [ad_mci_ev_p_value],\n",
    "                    \"Statistic\": [ad_mci_ev_statistic],\n",
    "                }\n",
    "            )\n",
    "            mci_healthy_ev_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [mci_healthy_ev_p_value],\n",
    "                    \"Statistic\": [mci_healthy_ev_statistic],\n",
    "                }\n",
    "            )\n",
    "            ad_healthy_ev_mannwhitneyu = pd.concat(\n",
    "                [ad_healthy_ev_mannwhitneyu, ad_healthy_ev_result_row], ignore_index=True\n",
    "            )\n",
    "            ad_mci_ev_mannwhitneyu = pd.concat(\n",
    "                [ad_mci_ev_mannwhitneyu, ad_mci_ev_result_row], ignore_index=True\n",
    "            )\n",
    "            mci_healthy_ev_mannwhitneyu = pd.concat(\n",
    "                [mci_healthy_ev_mannwhitneyu, mci_healthy_ev_result_row], ignore_index=True\n",
    "            )\n",
    "        healthy_assay_secreted = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 14\")\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 15\")\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "        mci_assay_secreted = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 14\")\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 15\")\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        ad_assay_secreted = (\n",
    "            (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 14\")\n",
    "            | (tidy_data_assay.index.get_level_values(\"Sample\") == \"SEC Fract 15\")\n",
    "        ) & (tidy_data_assay.index.get_level_values(\"Health\") == \"AD\")\n",
    "        healthy_assay_secreted_values = tidy_data_assay.loc[healthy_assay_secreted].values\n",
    "        mci_assay_secreted_values = tidy_data_assay.loc[mci_assay_secreted].values\n",
    "        ad_assay_secreted_values = tidy_data_assay.loc[ad_assay_secreted].values\n",
    "        if not (np.all(np.isnan(healthy_assay_secreted_values))) or not (np.all(np.isnan(mci_assay_secreted_values))) or not (np.all(np.isnan(ad_assay_secreted_values))):\n",
    "\n",
    "            ad_healthy_secreted_statistic, ad_healthy_secreted_p_value = mannwhitneyu(\n",
    "                ad_assay_secreted_values, healthy_assay_secreted_values\n",
    "            )\n",
    "            ad_mci_secreted_statistic, ad_mci_secreted_p_value = mannwhitneyu(\n",
    "                ad_assay_secreted_values, mci_assay_secreted_values\n",
    "            )\n",
    "            mci_healthy_secreted_statistic, mci_healthy_secreted_p_value = mannwhitneyu(\n",
    "                mci_assay_secreted_values, healthy_assay_secreted_values\n",
    "            )\n",
    "            ad_healthy_secreted_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [ad_healthy_secreted_p_value],\n",
    "                    \"Statistic\": [ad_healthy_secreted_statistic],\n",
    "                }\n",
    "            )\n",
    "            ad_mci_secreted_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [ad_mci_secreted_p_value],\n",
    "                    \"Statistic\": [ad_mci_secreted_statistic],\n",
    "                }\n",
    "            )\n",
    "            mci_healthy_secreted_result_row = pd.DataFrame(\n",
    "                {\n",
    "                    \"Assay\": [assay],\n",
    "                    \"p-value\": [mci_healthy_secreted_p_value],\n",
    "                    \"Statistic\": [mci_healthy_secreted_statistic],\n",
    "                }\n",
    "            )\n",
    "            ad_healthy_secreted_mannwhitneyu = pd.concat(\n",
    "                [ad_healthy_secreted_mannwhitneyu, ad_healthy_secreted_result_row],\n",
    "                ignore_index=True,\n",
    "            )\n",
    "            ad_mci_secreted_mannwhitneyu = pd.concat(\n",
    "                [ad_mci_secreted_mannwhitneyu, ad_mci_secreted_result_row], ignore_index=True\n",
    "            )\n",
    "            mci_healthy_secreted_mannwhitneyu = pd.concat(\n",
    "                [mci_healthy_secreted_mannwhitneyu, mci_healthy_secreted_result_row],\n",
    "                ignore_index=True,\n",
    "            )\n",
    "\n",
    "    ad_healthy_ev_sig_dif = ad_healthy_ev_mannwhitneyu[\n",
    "        ad_healthy_ev_mannwhitneyu[\"p-value\"] < 0.05\n",
    "    ]\n",
    "    ad_mci_ev_sig_dif = ad_mci_ev_mannwhitneyu[ad_mci_ev_mannwhitneyu[\"p-value\"] < 0.05]\n",
    "    mci_healthy_ev_sig_dif = mci_healthy_ev_mannwhitneyu[\n",
    "        mci_healthy_ev_mannwhitneyu[\"p-value\"] < 0.05\n",
    "    ]\n",
    "\n",
    "    ad_healthy_secreted_sig_dif = ad_healthy_secreted_mannwhitneyu[\n",
    "        ad_healthy_secreted_mannwhitneyu[\"p-value\"] < 0.05\n",
    "    ]\n",
    "    ad_mci_secreted_sig_dif = ad_mci_secreted_mannwhitneyu[\n",
    "        ad_mci_secreted_mannwhitneyu[\"p-value\"] < 0.05\n",
    "    ]\n",
    "    mci_healthy_secreted_sig_dif = mci_healthy_secreted_mannwhitneyu[\n",
    "        mci_healthy_secreted_mannwhitneyu[\"p-value\"] < 0.05\n",
    "    ]\n",
    "\n",
    "ad_healthy_ev_sig_dif_assays = set(ad_healthy_ev_sig_dif[\"Assay\"])\n",
    "ad_mci_ev_sig_dif_assays = set(ad_mci_ev_sig_dif[\"Assay\"])\n",
    "mci_healthy_ev_sig_dif_assays = set(mci_healthy_ev_sig_dif[\"Assay\"])\n",
    "\n",
    "ad_healthy_secreted_sig_dif_assays = set(ad_healthy_secreted_sig_dif[\"Assay\"])\n",
    "ad_mci_secreted_sig_dif_assays = set(ad_mci_secreted_sig_dif[\"Assay\"])\n",
    "mci_healthy_secreted_sig_dif_assays = set(mci_healthy_secreted_sig_dif[\"Assay\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a venn diagram of assays with significant differences in the EV fractions between sample groups\n",
    "venn3(\n",
    "    [\n",
    "        ad_healthy_ev_sig_dif_assays,\n",
    "        ad_mci_ev_sig_dif_assays,\n",
    "        mci_healthy_ev_sig_dif_assays,\n",
    "    ],\n",
    "    set_labels=(\"AD v. Healthy\", \"AD v. MCI\", \"MCI v. Healthy\"),\n",
    ")\n",
    "plt.title(\"Assays with Significant Differences in EV Fractions\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a venn diagram of assays with significant differences in the secreted fractions between sample groups\n",
    "venn3(\n",
    "    [\n",
    "        ad_healthy_secreted_sig_dif_assays,\n",
    "        mci_healthy_secreted_sig_dif_assays,\n",
    "        ad_mci_secreted_sig_dif_assays,\n",
    "    ],\n",
    "    set_labels=(\"AD v. Healthy\", \"MCI v. Healthy\", \"AD v. MCI\"),\n",
    ")\n",
    "plt.title(\"Assays with Significant Differences in Secreted Fractions\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a venn diagram of assays that have a significant difference between the sample groups\n",
    "venn2(\n",
    "    [\n",
    "        (ad_healthy_ev_sig_dif_assays & ad_healthy_secreted_sig_dif_assays),\n",
    "        (mci_healthy_ev_sig_dif_assays & mci_healthy_secreted_sig_dif_assays),\n",
    "    ],\n",
    "    set_labels=(\"AD v. Healthy\", \"Healthy v. MCI\"),\n",
    ")\n",
    "plt.title(\"Assays with Significant Differences Between Sample Groups\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fractionation analysis comparison of healthy vs. MCI spectrum samples\n",
    "healthy_ev_fractions = tidy_data[\n",
    "    (\n",
    "        (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 9\")\n",
    "        | (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 10\")\n",
    "    )\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "]\n",
    "\n",
    "healthy_fract6 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 6 \")\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "]\n",
    "\n",
    "healthy_fract11 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 11\")\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "]\n",
    "\n",
    "healthy_fract12 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 12\")\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "]\n",
    "\n",
    "healthy_fract13 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 13\")\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "]\n",
    "\n",
    "mci_ev_fractions = tidy_data[\n",
    "    (\n",
    "        (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 9\")\n",
    "        | (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 10\")\n",
    "    )\n",
    "    & (tidy_data.index.get_level_values(\"Health\") == \"MCI\")\n",
    "]\n",
    "\n",
    "mci_fract6 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 6 \")\n",
    "    & (\n",
    "        (tidy_data.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        | (tidy_data.index.get_level_values(\"Health\") == \"AD\")\n",
    "    )\n",
    "]\n",
    "\n",
    "mci_fract11 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 11\")\n",
    "    & (\n",
    "        (tidy_data.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        | (tidy_data.index.get_level_values(\"Health\") == \"AD\")\n",
    "    )\n",
    "]\n",
    "\n",
    "mci_fract12 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 12\")\n",
    "    & (\n",
    "        (tidy_data.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        | (tidy_data.index.get_level_values(\"Health\") == \"AD\")\n",
    "    )\n",
    "]\n",
    "\n",
    "mci_fract13 = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 13\")\n",
    "    & (\n",
    "        (tidy_data.index.get_level_values(\"Health\") == \"MCI\")\n",
    "        | (tidy_data.index.get_level_values(\"Health\") == \"AD\")\n",
    "    )\n",
    "]\n",
    "\n",
    "\n",
    "healthy_fractionation = []\n",
    "mci_fractionation = []\n",
    "\n",
    "\n",
    "for assay in list(tidy_data.columns):\n",
    "    if not tidy_data[assay].isna().all():\n",
    "        healthy_ev_fraction_median = healthy_ev_fractions[assay].median()\n",
    "        healthy_fract6_median = healthy_fract6[assay].median()\n",
    "        healthy_fract11_median = healthy_fract11[assay].median()\n",
    "        healthy_fract12_median = healthy_fract12[assay].median()\n",
    "        healthy_fract13_median = healthy_fract13[assay].median()\n",
    "        if (\n",
    "            (healthy_ev_fraction_median > healthy_fract6_median)\n",
    "            and (healthy_ev_fraction_median > healthy_fract11_median)\n",
    "            and (healthy_ev_fraction_median > healthy_fract12_median)\n",
    "            and (healthy_ev_fraction_median > healthy_fract13_median)\n",
    "        ):\n",
    "            healthy_fractionation.append(assay)\n",
    "        mci_ev_fraction_median = mci_ev_fractions[assay].median()\n",
    "        mci_fract6_median = mci_fract6[assay].median()\n",
    "        mci_fract11_median = mci_fract11[assay].median()\n",
    "        mci_fract12_median = mci_fract12[assay].median()\n",
    "        mci_fract13_median = mci_fract13[assay].median()\n",
    "        if (\n",
    "            (mci_ev_fraction_median > mci_fract6_median)\n",
    "            and (mci_ev_fraction_median > mci_fract11_median)\n",
    "            and (mci_ev_fraction_median > mci_fract12_median)\n",
    "            and (mci_ev_fraction_median > mci_fract13_median)\n",
    "        ):\n",
    "            mci_fractionation.append(assay)\n",
    "\n",
    "venn2(\n",
    "    [set(healthy_fractionation), set(mci_fractionation)],\n",
    "    set_labels=(\"Healthy\", \"MCI Spectrum\"),\n",
    ")\n",
    "plt.title(\"Samples with Correct Fractionation Pattern\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All samples fractionation analysis\n",
    "ev_fractions = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 9\")\n",
    "    | (tidy_data.index.get_level_values(\"Sample\") == \"SEC Fract 10\")\n",
    "]\n",
    "fract6 = tidy_data[(tidy_data.index.get_level_values(\"Sample\").str.contains(\"6\"))]\n",
    "fract11 = tidy_data[(tidy_data.index.get_level_values(\"Sample\").str.contains(\"11\"))]\n",
    "fract12 = tidy_data[(tidy_data.index.get_level_values(\"Sample\").str.contains(\"12\"))]\n",
    "fract13 = tidy_data[(tidy_data.index.get_level_values(\"Sample\").str.contains(\"13\"))]\n",
    "\n",
    "correct_fractionation = []\n",
    "annoying_assays = []\n",
    "\n",
    "for assay in list(tidy_data.columns):\n",
    "    if not tidy_data[assay].isna().all():\n",
    "        ev_fraction_median = ev_fractions[assay].median()\n",
    "        fract6_median = fract6[assay].median()\n",
    "        fract11_median = fract11[assay].median()\n",
    "        fract12_median = fract12[assay].median()\n",
    "        fract13_median = fract13[assay].median()\n",
    "        if (\n",
    "            (ev_fraction_median > fract6_median)\n",
    "            and (ev_fraction_median > fract11_median)\n",
    "            and (ev_fraction_median > fract12_median)\n",
    "            and (ev_fraction_median > fract13_median)\n",
    "        ):\n",
    "            correct_fractionation.append(assay)\n",
    "    else: \n",
    "        annoying_assays.append(assay)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(annoying_assays)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overlap between median assays, brain rna seq, and internal only tmhmm\n",
    "\n",
    "# Internal microglia\n",
    "internal_microglia_targets = (\n",
    "    set(correct_fractionation) & set(microglia_uniprot_ids) & set(inside_uniprots)\n",
    ")\n",
    "\n",
    "# Internal astrocyte\n",
    "internal_astrocyte_targets = (\n",
    "    set(correct_fractionation) & set(astrocyte_uniprot_ids) & set(inside_uniprots)\n",
    ")\n",
    "\n",
    "# Internal oligodendrocyte\n",
    "internal_oligodendrocyte_targets = (\n",
    "    set(correct_fractionation) & set(oligodendrocyte_uniprot_ids) & set(inside_uniprots)\n",
    ")\n",
    "\n",
    "# Internal neuron\n",
    "internal_neuron_targets = (\n",
    "    set(correct_fractionation) & set(neuron_uniprot_ids) & set(inside_uniprots)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tm_peak = set(tm_uniprots) & set(correct_fractionation)\n",
    "internal_peak = set(inside_uniprots) & set(correct_fractionation)\n",
    "external_peak = set(outside_uniprots) & set(correct_fractionation)\n",
    "\n",
    "peak_dict = {\n",
    "    \"Transmembrane\": len(tm_peak),\n",
    "    \"Internal\": len(internal_peak),\n",
    "    \"External\": len(external_peak),\n",
    "}\n",
    "\n",
    "x_labels = list(peak_dict.keys())\n",
    "y_labels = list(peak_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "plt.xlabel(\"Target Localization\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Localization of Targets with EV Peaks\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "internal_with_peak = set(inside_uniprots) & set(correct_fractionation)\n",
    "internal_without_peak = set(inside_uniprots) - set(correct_fractionation)\n",
    "\n",
    "int_dict = {\n",
    "    \"Targets with Peak\": len(internal_with_peak),\n",
    "    \"Targets without Peak\": len(internal_without_peak),\n",
    "}\n",
    "\n",
    "x_labels = list(int_dict.keys())\n",
    "y_labels = list(int_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "plt.xlabel(\"Internal Targets\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Internal Targets with and without EV Peak\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "external_with_peak = set(outside_uniprots) & set(correct_fractionation)\n",
    "external_without_peak = set(outside_uniprots) - set(correct_fractionation)\n",
    "\n",
    "ext_dict = {\n",
    "    \"Targets with Peak\": len(external_with_peak),\n",
    "    \"Targets Without Peak\": len(external_without_peak),\n",
    "}\n",
    "\n",
    "x_labels = list(ext_dict.keys())\n",
    "y_labels = list(ext_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "plt.xlabel(\"Transmembrane Targets\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"External Targets with and without EV Peak\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tm_with_peak = set(tm_uniprots) & set(correct_fractionation)\n",
    "tm_without_peak = set(tm_uniprots) - set(correct_fractionation)\n",
    "\n",
    "tm_dict = {\n",
    "    \"Targets with Peak\": len(tm_with_peak),\n",
    "    \"Targets without Peak\": len(tm_without_peak),\n",
    "}\n",
    "\n",
    "x_labels = list(tm_dict.keys())\n",
    "y_labels = list(tm_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "plt.xlabel(\"Transmembrane Targets\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Transmembrane Targets with and without EV Peak\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overlap between median assays, brain rna seq, and transmembrane targets\n",
    "\n",
    "# TM microglia\n",
    "tm_microglia_targets = (\n",
    "    set(correct_fractionation) & set(microglia_uniprot_ids) & set(tm_uniprots)\n",
    ")\n",
    "\n",
    "# TM astrocyte\n",
    "tm_astrocyte_targets = (\n",
    "    set(correct_fractionation) & set(astrocyte_uniprot_ids) & set(tm_uniprots)\n",
    ")\n",
    "\n",
    "# TM oligodendrocyte\n",
    "tm_oligodendrocyte_targets = (\n",
    "    set(correct_fractionation) & set(oligodendrocyte_uniprot_ids) & set(tm_uniprots)\n",
    ")\n",
    "\n",
    "# TM neuron\n",
    "tm_neuron_targets = (\n",
    "    set(correct_fractionation) & set(neuron_uniprot_ids) & set(tm_uniprots)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of transmembrane targets per cell type\n",
    "\n",
    "internal_dict = {\n",
    "    \"Microglia\": len(tm_microglia_targets),\n",
    "    \"Astrocytes\": len(tm_astrocyte_targets),\n",
    "    \"Oligodendrocytes\": len(tm_oligodendrocyte_targets),\n",
    "    \"Neurons\": len(tm_neuron_targets),\n",
    "}\n",
    "\n",
    "x_labels = list(internal_dict.keys())\n",
    "y_labels = list(internal_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "\n",
    "for i, v in enumerate(y_labels):\n",
    "    plt.text(i, v + 0.5, str(v), ha=\"center\", va=\"bottom\")\n",
    "\n",
    "plt.yticks([1, 2, 3, 4, 5, 6, 7, 8])\n",
    "\n",
    "plt.xlabel(\"Cell Type\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Transmembrane Targets for Each Cell Type\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "internal_dict = {\n",
    "    \"Microglia\": len(internal_microglia_targets),\n",
    "    \"Astrocytes\": len(internal_astrocyte_targets),\n",
    "    \"Oligodendrocytes\": len(internal_oligodendrocyte_targets),\n",
    "    \"Neurons\": len(internal_neuron_targets),\n",
    "}\n",
    "\n",
    "x_labels = list(internal_dict.keys())\n",
    "y_labels = list(internal_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "\n",
    "for i, v in enumerate(y_labels):\n",
    "    plt.text(i, v + 0.5, str(v), ha=\"center\", va=\"bottom\")\n",
    "\n",
    "plt.yticks([1, 2, 3, 4, 5, 6])\n",
    "\n",
    "plt.xlabel(\"Cell Type\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Internal Targets for Each Cell Type\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CSF_SAMPLES = [\n",
    "    \"SEC Fract 6 \",\n",
    "    \"SEC Fract 7\",\n",
    "    \"SEC Fract 8\",\n",
    "    \"SEC Fract 9\",\n",
    "    \"SEC Fract 10\",\n",
    "    \"SEC Fract 11\",\n",
    "    \"SEC Fract 12\",\n",
    "    \"SEC Fract 13\",\n",
    "    \"SEC Fract 14\",\n",
    "    \"SEC Fract 15\",\n",
    "    \"CSF A Internal EV\",\n",
    "    \"CSF B Internal EV\",\n",
    "    \"CSF C Internal EV\",\n",
    "    \"CSF D Internal EV\",\n",
    "]\n",
    "\n",
    "def graph_medians(uniprot_id):\n",
    "    df = tidy_data[uniprot_id]\n",
    "    df = df.reset_index(level=[\"SampleID\", \"Health\", \"Sample\"])\n",
    "\n",
    "    df[\"Sample\"] = pd.Categorical(df[\"Sample\"], categories=CSF_SAMPLES, ordered=True)\n",
    "\n",
    "    df_sorted = df.sort_values(\"Sample\")\n",
    "\n",
    "    grouped_data = [\n",
    "        group[uniprot_id].values for name, group in df_sorted.groupby(\"Sample\")\n",
    "    ]\n",
    "\n",
    "    plt.boxplot(grouped_data, notch=None, vert=None, patch_artist=None, widths=None)\n",
    "\n",
    "    plt.xlabel(\"Sample Description\")\n",
    "\n",
    "    plt.ylabel(\"Delta\")\n",
    "\n",
    "    plt.title(f\"{uniprot_id} Fractionation Pattern\")\n",
    "\n",
    "    plt.xticks(range(1, len(CSF_SAMPLES) + 1), CSF_SAMPLES)\n",
    "\n",
    "    plt.xticks(rotation=45, ha=\"right\")\n",
    "\n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
