{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    \"C:\\\\Users\\\\Wyss User\\\\AppData\\\\Local\\\\Packages\\\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\\\LocalCache\\\\local-packages\\\\Python311\\\\site-packages\"\n",
    ")\n",
    "\n",
    "import biolib\n",
    "import gzip\n",
    "import math\n",
    "import os\n",
    "import re\n",
    "import requests\n",
    "import pathlib\n",
    "import tspex\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests as r\n",
    "import seaborn as sns\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from IPython.display import Image\n",
    "from matplotlib_venn import venn2\n",
    "from matplotlib_venn import venn3\n",
    "from scipy import stats\n",
    "from scipy.stats import f_oneway\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ASSAY_LIST_PATH = \"assay-list-olink-explore-3072.xlsx\"\n",
    "BRAIN_RNA_SEQ_PATH = \"fe-wp-dataset-124.csv\"\n",
    "HGNC_IDS = (\n",
    "    \"https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt\"\n",
    ")\n",
    "PLATE_LAYOUT_PATH = \"Plate Layout.xlsx\"\n",
    "RAW_DATA_FILE = \"LCSET_28343_10-4-2023_EXTENDED_NPX_2023-10-20.csv\"\n",
    "UNIPROT_TO_FASTA_PATH = \"uniprotkb_proteome_UP000005640_2023_11_20.fasta.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "assays = pd.read_excel(ASSAY_LIST_PATH)\n",
    "\n",
    "# uniprot_ids = []\n",
    "# fasta_sequences = []\n",
    "\n",
    "# assay_ids = list(assays[\"UniProt ID\"])\n",
    "# uniprot_ids = []\n",
    "# fasta_sequences = []\n",
    "# for cID in assay_ids:\n",
    "#     baseUrl=\"http://www.uniprot.org/uniprot/\"\n",
    "#     currentUrl=baseUrl+cID+\".fasta\"\n",
    "#     response = r.post(currentUrl)\n",
    "#     cData=''.join(response.text)\n",
    "#     uniprot_id = re.search(r'\\|([^|]+)\\|', cData)\n",
    "#     if match:\n",
    "#         uniprot_ids.append(uniprot_id.group(1))\n",
    "#     fasta = cData.split('\\n', 1)\n",
    "#     if len(fasta) > 1:\n",
    "#         fasta_sequences.extend(fasta[1].split('|'))\n",
    "# formatted_fasta_sequences = [item.replace('\\n', '') for item in fasta_sequences]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "brain_rna_seq = pd.read_csv(BRAIN_RNA_SEQ_PATH)\n",
    "\n",
    "hgnc_uniprot_mapping_data = pd.read_csv(\n",
    "    (StringIO(requests.get(HGNC_IDS).text)),\n",
    "    sep=\"\\t\",\n",
    "    usecols=[\"hgnc_id\", \"uniprot_ids\"],\n",
    ")\n",
    "brain_rna_seq = pd.merge(\n",
    "    brain_rna_seq,\n",
    "    hgnc_uniprot_mapping_data,\n",
    "    left_on=\"id\",\n",
    "    right_on=\"hgnc_id\",\n",
    "    how=\"inner\",\n",
    ")\n",
    "brain_rna_seq.dropna(subset=[\"uniprot_ids\"], inplace=True)\n",
    "brain_rna_seq.drop_duplicates(subset=[\"uniprot_ids\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map UniProt IDs to the corresponding FASTA sequences\n",
    "def parse_gz_file(file_path):\n",
    "    protein_dict = {}\n",
    "    current_uniprot_id = None\n",
    "    current_sequence = \"\"\n",
    "    with gzip.open(file_path, \"rt\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if current_uniprot_id is not None:\n",
    "                    protein_dict[current_uniprot_id] = current_sequence\n",
    "                    current_sequence = \"\"\n",
    "                if \"|\" in line:\n",
    "                    current_uniprot_id = line.split(\"|\")[1].strip()\n",
    "                else:\n",
    "                    print(f\"Skipping line without expected format: {line}\")\n",
    "                    current_uniprot_id = None\n",
    "            else:\n",
    "                current_sequence += line\n",
    "        if current_uniprot_id is not None:\n",
    "            protein_dict[current_uniprot_id] = current_sequence\n",
    "    return protein_dict\n",
    "\n",
    "\n",
    "fasta_sequences = parse_gz_file(UNIPROT_TO_FASTA_PATH)\n",
    "\n",
    "assays[\"Sequence\"] = assays[\"UniProt ID\"].map(lambda x: fasta_sequences.get(x, \"N/A\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tmhmm_localization(assays, output_directory):\n",
    "    deeptmhmm = biolib.load(\"DTU/DeepTMHMM\")\n",
    "    with open(\"query.fasta\", \"w\") as fasta_file:\n",
    "        for _, row in assays.iterrows():\n",
    "            fasta_line = f\">{row['UniProt ID']}\\n{row['Sequence']}\\n\"\n",
    "            fasta_file.write(fasta_line)\n",
    "        deeptmhmm_job = deeptmhmm.cli(args=\"--fasta query.fasta\", machine=\"local\")\n",
    "        deeptmhmm_job.save_files(output_directory)\n",
    "\n",
    "\n",
    "def identify_localization(assays, region, output_directory=\"olink_output\"):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    assays : pandas.DataFrame\n",
    "        DataFrame with columns called 'UniProt ID' and 'Sequence'\n",
    "    region : {'TMhelix', 'inside', 'outside', 'internal'}\n",
    "        Subcellular region requested. Options:\n",
    "          - 'TMhelix': transmembrane proteins\n",
    "          - 'inside': at least some of the protein is inside the cell/EV\n",
    "          - 'outside': at least some of the protein is outside the cell/EV\n",
    "          - 'internal': the protein is only found inside the cell, no transmembrane or outside domains\n",
    "          - 'external': the protein is only found outside the cell, no transmembrane or inside domains\n",
    "    \"\"\"\n",
    "    output_directory_path = pathlib.Path(output_directory)\n",
    "    if not os.path.exists(output_directory_path):\n",
    "        os.makedirs(output_directory_path)\n",
    "    # os.chdir(output_directory)\n",
    "    if not os.path.exists(output_directory_path / \"TMRs.gff3\"):\n",
    "        tmhmm_localization(assays, output_directory_path)\n",
    "    localization_df = pd.read_csv(\n",
    "        output_directory_path / \"TMRs.gff3\",\n",
    "        sep=\"\\t\",\n",
    "        comment=\"#\",\n",
    "        names=[\n",
    "            \"uniprot_id\",\n",
    "            \"region_location\",\n",
    "            \"region_start\",\n",
    "            \"region_end\",\n",
    "            0,\n",
    "            1,\n",
    "            2,\n",
    "            3,\n",
    "        ],\n",
    "    )\n",
    "    localization_df = localization_df[localization_df[\"uniprot_id\"] != \"//\"].dropna(\n",
    "        axis=1\n",
    "    )\n",
    "    get_regional_uniprots = lambda region: set(\n",
    "        localization_df[localization_df[\"region_location\"] == region][\"uniprot_id\"]\n",
    "    )\n",
    "    if region == \"internal\":\n",
    "        tm_uniprots = get_regional_uniprots(\"TMhelix\")\n",
    "        outside_uniprots = get_regional_uniprots(\"outside\")\n",
    "        inside_uniprots = get_regional_uniprots(\"inside\")\n",
    "        return inside_uniprots - tm_uniprots - outside_uniprots\n",
    "    if region == \"external\":\n",
    "        tm_uniprots = get_regional_uniprots(\"TMhelix\")\n",
    "        outside_uniprots = get_regional_uniprots(\"outside\")\n",
    "        inside_uniprots = get_regional_uniprots(\"inside\")\n",
    "        return outside_uniprots - inside_uniprots - tm_uniprots\n",
    "    else:\n",
    "        return get_regional_uniprots(region)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_mean(df):\n",
    "    return df.assign(Mean=df.mean(axis=1, numeric_only=True))\n",
    "\n",
    "\n",
    "def mean_cell_type(brain_rna_seq_data, cell_type):\n",
    "    \"\"\"\n",
    "    Returns only the mean of the data for the specified cell type, as well as the UniProt ID information in an additional column\n",
    "    Parameters\n",
    "    ----------\n",
    "    brain_rna_seq_data : pandas.DataFrame\n",
    "        Dataframe with a column called \"uniprot_ids\" (contains UniProt ID), and other columns containing cell-type specific Brain RNA Seq data\n",
    "    cell_type : {'fetal', 'mature', 'endothelial', 'microglia', 'oligodendrocyte', 'neuron'}\n",
    "        Cell type of interest requested. Options:\n",
    "         - 'astrocyte': mature astrocytes\n",
    "         - 'endothelial': endothelial cells\n",
    "         - 'microglia': microglia cells\n",
    "         - 'oligodendrocyte': oligodendrocytes\n",
    "         - 'neuron': neurons\n",
    "    \"\"\"\n",
    "    if cell_type == \"microglia\":\n",
    "        microglia_df = calculate_mean(\n",
    "            brain_rna_seq_data[\n",
    "                brain_rna_seq_data.filter(like=\"microglla\").columns.append(\n",
    "                    pd.Index([\"uniprot_ids\"])\n",
    "                )\n",
    "            ]\n",
    "        )[[\"uniprot_ids\", \"Mean\"]]\n",
    "        return microglia_df.rename(\n",
    "            columns={\"uniprot_ids\": \"uniprot_ids\", \"Mean\": \"microglia\"}\n",
    "        )\n",
    "    if cell_type == \"astrocyte\":\n",
    "        astrocyte_df = calculate_mean(\n",
    "            brain_rna_seq_data[\n",
    "                brain_rna_seq_data.filter(like=\"mature\").columns.append(\n",
    "                    pd.Index([\"uniprot_ids\"])\n",
    "                )\n",
    "            ]\n",
    "        )[[\"uniprot_ids\", \"Mean\"]]\n",
    "        return astrocyte_df.rename(\n",
    "            columns={\"uniprot_ids\": \"uniprot_ids\", \"Mean\": \"astrocyte\"}\n",
    "        )\n",
    "    else:\n",
    "        cell_type_df = calculate_mean(\n",
    "            brain_rna_seq_data[\n",
    "                brain_rna_seq_data.filter(like=cell_type).columns.append(\n",
    "                    pd.Index([\"uniprot_ids\"])\n",
    "                )\n",
    "            ]\n",
    "        )[[\"uniprot_ids\", \"Mean\"]]\n",
    "        return cell_type_df.rename(\n",
    "            columns={\"uniprot_ids\": \"uniprot_ids\", \"Mean\": cell_type}\n",
    "        )\n",
    "\n",
    "\n",
    "def cell_type_enrichment(\n",
    "    brain_rna_seq_data,\n",
    "    cell_type,\n",
    "    indiv_specificity_metric,\n",
    "    indiv_specificity_cutoff,\n",
    "    general_specificity_metric=\"none\",\n",
    "    general_specificity_cutoff=\"none\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    brain_rna_seq_data : pandas.DataFrame\n",
    "        Dataframe with a column called \"uniprot_ids\" (contains UniProt ID), and other columns containing cell-type specific Brain RNA Seq data for the cell types listed under cell_type\n",
    "    cell_type : {'fetal', 'mature', 'endothelial', 'microglia', 'oligodendrocyte', 'neuron'}\n",
    "        Cell type of interest requested. Options:\n",
    "         - 'astrocyte': mature astrocytes\n",
    "         - 'endothelial': endothelial cells\n",
    "         - 'microglia': microglia cells\n",
    "         - 'oligodendrocyte': oligodendrocytes\n",
    "         - 'neuron': neurons\n",
    "    general_specificity_metric : {\n",
    "        'tau',\n",
    "        'counts',\n",
    "        'gini',\n",
    "        'simpson',\n",
    "        'shannon_specifity',\n",
    "        'roku_specificity',\n",
    "        'spm_dpm',\n",
    "        'js_specificity_dpm',\n",
    "    }\n",
    "        Primary metric of determining cell-type specificity requested. Options:\n",
    "        - 'tau': tau score\n",
    "        - 'counts': counts\n",
    "        - 'gini': gini coefficient\n",
    "        - 'simpson': simpson index\n",
    "        - 'shannon_specificity': shannon entropy specificity (HS)\n",
    "        - 'roku_specificity': ROKU specificity\n",
    "        - 'spm_dpm': specificity measure dispersion (SPM DPM)\n",
    "        - 'js_specificity_dpm': jensen-shannon specificity dispersion (JSS DPM)\n",
    "        - 'none': default, no general specificity metric requested\n",
    "    indiv_specificity_metric: {\n",
    "        'tsi',\n",
    "        'zscore',\n",
    "        'spm',\n",
    "        'js_specificity'\n",
    "    }\n",
    "        Secondary metric of determining cell type specificity requested. Options:\n",
    "        - 'tsi': tissue specificity index\n",
    "        - 'z-score': z-score\n",
    "        - 'spm': specificity measure\n",
    "        - 'js_specificity': jensen-shannon specificity\n",
    "    general_specificity_cutoff : numeric\n",
    "        Numeric value representing the minimum value of the first enrichment cutoff.\n",
    "        Default = none\n",
    "    indiv_specificity_cutoff : numeric\n",
    "        Numeric value representing the minimum value of the second enrichment cutoff.\n",
    "    \"\"\"\n",
    "    astrocytes = mean_cell_type(brain_rna_seq_data, \"astrocyte\")\n",
    "    endothelial = mean_cell_type(brain_rna_seq_data, \"endothelial\")\n",
    "    microglia = mean_cell_type(brain_rna_seq_data, \"microglia\")\n",
    "    oligodendrocytes = mean_cell_type(brain_rna_seq_data, \"oligodendrocyte\")\n",
    "    neurons = mean_cell_type(brain_rna_seq_data, \"neuron\")\n",
    "\n",
    "    all_cell_types = pd.merge(\n",
    "        pd.merge(\n",
    "            pd.merge(\n",
    "                pd.merge(astrocytes, endothelial, on=\"uniprot_ids\"),\n",
    "                microglia,\n",
    "                on=\"uniprot_ids\",\n",
    "            ),\n",
    "            oligodendrocytes,\n",
    "            on=\"uniprot_ids\",\n",
    "        ),\n",
    "        neurons,\n",
    "        on=\"uniprot_ids\",\n",
    "    )\n",
    "    all_cell_types.set_index(\"uniprot_ids\", inplace=True)\n",
    "\n",
    "    expression_data = all_cell_types.loc[\n",
    "        :, [\"endothelial\", \"astrocyte\", \"microglia\", \"neuron\", \"oligodendrocyte\"]\n",
    "    ]\n",
    "    expression_data = expression_data.loc[(expression_data > 0).any(axis=1)]\n",
    "\n",
    "    if general_specificity_metric == \"none\":\n",
    "        tso_indiv = tspex.TissueSpecificity(\n",
    "            expression_data, indiv_specificity_metric, log=True\n",
    "        )\n",
    "    else:\n",
    "        tso_gen = tspex.TissueSpecificity(\n",
    "            expression_data, general_specificity_metric, log=True\n",
    "        )\n",
    "        tso_gen_df = pd.DataFrame(\n",
    "            tso_gen.tissue_specificity, columns=[\"general_specificity_metric\"]\n",
    "        )\n",
    "        tso_gen_filtered = list(\n",
    "            tso_gen_df[\n",
    "                tso_gen_df[\"general_specificity_metric\"] > general_specificity_cutoff\n",
    "            ].index\n",
    "        )\n",
    "        tso_gen_filtered_df = expression_data.loc[tso_gen_filtered]\n",
    "\n",
    "        tso_indiv = tspex.TissueSpecificity(\n",
    "            tso_gen_filtered_df, indiv_specificity_metric, log=True\n",
    "        )\n",
    "\n",
    "    tso_indiv_df = pd.DataFrame(\n",
    "        tso_indiv.tissue_specificity,\n",
    "        columns=[\"endothelial\", \"astrocyte\", \"microglia\", \"neuron\", \"oligodendrocyte\"],\n",
    "    )\n",
    "    cell_type_uniprot_ids = tso_indiv_df[\n",
    "        tso_indiv_df[cell_type] > indiv_specificity_cutoff\n",
    "    ].index.tolist()\n",
    "    return cell_type_uniprot_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIMPORTANT_COLUMNS = [\n",
    "    \"Index\",\n",
    "    \"OlinkID\",\n",
    "    \"Panel_Lot_Nr\",\n",
    "    \"PlateID\",\n",
    "    \"WellID\",\n",
    "    \"Normalization\",\n",
    "    \"Processing_StartDate\",\n",
    "    \"Processing_EndDate\",\n",
    "    \"AnalyzerID\",\n",
    "    \"ExploreVersion\",\n",
    "]\n",
    "\n",
    "PANELS = [\n",
    "    \"Cardiometabolic\",\n",
    "    \"Cardiometabolic_II\",\n",
    "    \"Inflammation\",\n",
    "    \"Inflammation_II\",\n",
    "    \"Neurology\",\n",
    "    \"Neurology_II\",\n",
    "    \"Oncology\",\n",
    "    \"Oncology_II\",\n",
    "]\n",
    "\n",
    "# Clean up the raw OLINK data file and map the sample locations to their barcodes\n",
    "plate_layout = pd.read_excel(PLATE_LAYOUT_PATH, index_col=0)\n",
    "raw_data = pd.read_csv(RAW_DATA_FILE, sep=\";\")\n",
    "raw_data = raw_data.drop(UNIMPORTANT_COLUMNS, axis=1)\n",
    "\n",
    "data = pd.merge(raw_data, plate_layout, how=\"left\", on=\"SampleID\")\n",
    "\n",
    "# Normalize the data according to plate negative controls\n",
    "ctrl_dict = {}\n",
    "for panel in PANELS:\n",
    "    df = data[(data[\"Panel\"] == panel)]\n",
    "    plate_ctrl = df[\n",
    "        (df[\"Sample_Type\"] == \"PLATE_CONTROL\")\n",
    "        & (df[\"Assay\"] == \"Amplification control 1\")\n",
    "    ][\"NPX\"].median()\n",
    "    ctrl_dict[panel] = plate_ctrl\n",
    "data[\"Delta\"] = data.apply(\n",
    "    lambda row: row[\"NPX\"] - ctrl_dict[row[\"Panel\"]],\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "# Linearize the NPX values\n",
    "data[\"Linear Delta\"] = data[\"Delta\"].map(lambda x: 2**x)\n",
    "\n",
    "# Identify and drop replicate assays\n",
    "vc = data[[\"SampleID\", \"Sample_Type\", \"UniProt\"]].value_counts()\n",
    "vc = vc[vc > 1]\n",
    "\n",
    "unique_uniprot_values = vc.index.get_level_values(\"UniProt\").unique()\n",
    "uniprot_list = unique_uniprot_values.tolist()\n",
    "\n",
    "raw_data_no_reps = data[~data[\"UniProt\"].isin(uniprot_list)].copy()\n",
    "for index, row in raw_data_no_reps.iterrows():\n",
    "    if row[\"NPX\"] < row[\"LOD\"]:\n",
    "        raw_data_no_reps.at[index, \"NPX\"] = row[\"LOD\"]\n",
    "\n",
    "# Create a tidy dataframe for the delta values\n",
    "tidy_data = raw_data_no_reps[raw_data_no_reps[\"Sample_Type\"] == \"SAMPLE\"].pivot(\n",
    "    columns=\"UniProt\", index=[\"SampleID\", \"Health\", \"Sample\", \"CSF_sample\"], values=\"Linear Delta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_fractionation(tidy_dataframe, high_fractions, low_fractions, sample_health = \"all\", mean_median_individual = \"individual\"):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    'tidy_dataframe' : pandas.DataFrame\n",
    "        DataFrame with one column for each assay, one row for each sample, linearized NPX as the vlaues, and the following indices:\n",
    "            - 'SampleID'\n",
    "            - 'Health'\n",
    "            - 'Sample'\n",
    "            - 'CSF_sample'\n",
    "    'high_fractions' : list of strings\n",
    "        Fractions that should be higher than the fractions in the list of low fractions.\n",
    "    'low_fractions' : list of strings\n",
    "        Fractions that should be lower than the fractions in the list of high fractions.\n",
    "    'sample_health' : {'all', 'ad', 'mci', 'mci_spectrum'}\n",
    "        Health of the sample requested. Options: \n",
    "            - 'healthy': only samples from healthy individuals\n",
    "            - 'all': all different health groups\n",
    "            - 'ad': samples from individuals diagnosed with Alzheimer's Disease (AD)\n",
    "            - 'mci': samples from individuals diagnosed with mild cognitive imapirment that has not yet progressed to AD\n",
    "            - 'mci_spectrum': samples from individuals diagnosed with mild cognitive impairment and samples from individuals that have been diagnosed with AD\n",
    "    'mean_median_individual' : {'mean', 'median', 'individual'}\n",
    "        How the groups of samples should be analyzed. Options: \n",
    "            - 'mean': the means of each fraction should be compared against each other\n",
    "            - 'median': the medians of each fraction should be compared against each other\n",
    "            - 'individual': the fractions of each sample should be compared against each other with no aggregation/grouping \n",
    "            - 'individual_median': for each fraction, the median value of all samples will be compared\n",
    "            - 'individual_mean': for each fraction, the mean value of all samples will be compared\n",
    "        Default value: 'individual'\n",
    "    \"\"\"\n",
    "    \n",
    "    non_ppa_data = tidy_dataframe[tidy_dataframe.index.get_level_values(\"Sample\").str.contains(\"SEC\")]\n",
    "\n",
    "    if sample_health == 'healthy': \n",
    "        requested_health_data = non_ppa_data[non_ppa_data.index.get_level_values(\"Health\").str.contains('Healthy')]\n",
    "    if sample_health == 'all': \n",
    "        requested_health_data = non_ppa_data\n",
    "    if sample_health == 'ad': \n",
    "        requested_health_data =  non_ppa_data[non_ppa_dat.index.get_level_values(\"Health\").str.contains('AD')]\n",
    "    if sample_health == 'mci':\n",
    "        requested_health_data =  non_ppa_data[non_ppa_dat.index.get_level_values(\"Health\").str.contains('MCI')]\n",
    "    if sample_health == 'mci_spectrum': \n",
    "        requested_health_data = non_ppa_data[\n",
    "            (non_ppa_data.index.get_level_values(\"Health\").str.contains('AD'))\n",
    "            | (non_ppa_data.index.get_level_values('Health').str.contains('MCI'))\n",
    "        ]\n",
    "\n",
    "    high_fractions_dataframes = {}\n",
    "    for fraction in high_fractions:\n",
    "        high_fractions_dataframes[fraction] = requested_health_data[requested_health_data.index.get_level_values(\"Sample\").str.contains(fraction)]\n",
    "    high_fractions_df = pd.concat(high_fractions_dataframes.values(), axis = 0)\n",
    "    low_fractions_dataframes = {}\n",
    "    for fraction in low_fractions: \n",
    "        low_fractions_dataframes[fraction] = requested_health_data[requested_health_data.index.get_level_values(\"Sample\").str.contains(fraction)]\n",
    "    low_fractions_df = pd.concat(low_fractions_dataframes.values(), axis=0)\n",
    "\n",
    "    correct_fractionation = []\n",
    "    fractionation = []\n",
    "    for assay in list(non_ppa_data.columns):\n",
    "        if mean_median_individual == \"median\":\n",
    "            if not high_fractions_df[assay].isna().all():\n",
    "                high_fractions = high_fractions_df[assay].median()\n",
    "            if not low_fractions_df[assay].isna().all():\n",
    "                low_fractions = low_fractions_df[assay].median()\n",
    "            if high_fractions > low_fractions: \n",
    "                correct_fractionation.append(assay)\n",
    "        if mean_median_individual == \"mean\":\n",
    "            if not high_fractions_df[assay].isna().all():\n",
    "                high_fractions = high_fractions_df[assay].mean()\n",
    "            if not low_fractions_df[assay].isna().all():\n",
    "                low_fractions = low_fractions_df[assay].mean()\n",
    "            if high_fractions > low_fractions: \n",
    "                correct_fractionation.append(assay)\n",
    "        if mean_median_individual == \"individual\":\n",
    "            for sample in list(requested_health_data.index.get_level_values(\"CSF_sample\").unique()):\n",
    "                high_sample_data = high_fractions_df[high_fractions_df.index.get_level_values(\"CSF_sample\") == sample]\n",
    "                low_sample_data = low_fractions_df[low_fractions_df.index.get_level_values(\"CSF_sample\") == sample]\n",
    "                if not high_sample_data[assay].isna().all():\n",
    "                    high_fractions = high_sample_data[assay].min()\n",
    "                if not low_sample_data[assay].isna().all():\n",
    "                    low_fractions = low_sample_data[assay].max()\n",
    "                if high_fractions > low_fractions: \n",
    "                    correct_fractionation.append(assay)\n",
    "        if mean_median_individual == \"individual_median\":\n",
    "            high_fractions_values = []\n",
    "            for fraction in high_fractions:\n",
    "                if not high_fractions_df[assay].isna().all():\n",
    "                    high_fractions_values.append(high_fractions_df[high_fractions_df.index.get_level_values(\"Sample\").str.contains(fraction)][assay].median())\n",
    "            low_fractions_values = []\n",
    "            for fraction in low_fractions: \n",
    "                if not low_fractions_df[assay].isna().all():\n",
    "                    low_fractions_values.append(low_fractions_df[low_fractions_df.index.get_level_values(\"Sample\").str.contains(fraction)][assay].median())\n",
    "            if all(value1 > value2 for value1 in high_fractions_values for value2 in low_fractions_values):\n",
    "                correct_fractionation.append(assay)\n",
    "        if mean_median_individual == \"individual_mean\":\n",
    "            high_fractions_values = []\n",
    "            for fraction in high_fractions:\n",
    "                if not high_fractions_df[assay].isna().all():\n",
    "                    high_fractions_values.append(high_fractions_df[high_fractions_df.index.get_level_values(\"Sample\").str.contains(fraction)][assay].mean())\n",
    "            low_fractions_values = []\n",
    "            for fraction in low_fractions: \n",
    "                if not low_fractions_df[assay].isna().all():\n",
    "                    low_fractions_values.append(low_fractions_df[low_fractions_df.index.get_level_values(\"Sample\").str.contains(fraction)][assay].mean())\n",
    "            if all(value1 > value2 for value1 in high_fractions_values for value2 in low_fractions_values):\n",
    "                correct_fractionation.append(assay)\n",
    "    # return pd.Series(correct_fractionation).value_counts()\n",
    "    return correct_fractionation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_targets(correct_fractionation_uniprot_ids, cell_type_uniprot_ids, localization_uniprot_ids):\n",
    "    return (set(correct_fractionation) & set(cell_type_uniprot_ids) & set(localization_uniprot_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'correct_fractionation' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\Wyss User\\Documents\\EVs\\OLINK\\10_analysis.ipynb Cell 11\u001b[0m line \u001b[0;36m3\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m     correct_fractionation_uniprot_ids \u001b[39m=\u001b[39m analyze_fractionation(tidy_dataframe, high_fractions, low_fractions, sample_health, mean_median_individual)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m identify_targets(correct_fractionation_uniprot_ids, cell_type_uniprot_ids, localization_uniprot_ids)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=31'>32</a>\u001b[0m identify_fractionation_pattern_cell_type_specific_targets(assays, UNIPROT_TO_FASTA_PATH, \u001b[39m\"\u001b[39;49m\u001b[39minternal\u001b[39;49m\u001b[39m\"\u001b[39;49m, brain_rna_seq, \u001b[39m\"\u001b[39;49m\u001b[39mmicroglia\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mtau\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m0.7\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mzscore\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m0.45\u001b[39;49m, tidy_data, [\u001b[39m'\u001b[39;49m\u001b[39m9\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39m10\u001b[39;49m\u001b[39m'\u001b[39;49m], [\u001b[39m'\u001b[39;49m\u001b[39m6\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39m11\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39m12\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39m13\u001b[39;49m\u001b[39m'\u001b[39;49m], \u001b[39m'\u001b[39;49m\u001b[39mhealthy\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mindividual_median\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "\u001b[1;32mc:\\Users\\Wyss User\\Documents\\EVs\\OLINK\\10_analysis.ipynb Cell 11\u001b[0m line \u001b[0;36m3\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m cell_type_uniprot_ids \u001b[39m=\u001b[39m cell_type_enrichment(\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=24'>25</a>\u001b[0m     brain_rna_seq_dataframe, cell_type, indiv_specificity_metric, indiv_specificity_cutoff, general_specificity_metric, general_specificity_cutoff\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m )\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m correct_fractionation_uniprot_ids \u001b[39m=\u001b[39m analyze_fractionation(tidy_dataframe, high_fractions, low_fractions, sample_health, mean_median_individual)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m \u001b[39mreturn\u001b[39;00m identify_targets(correct_fractionation_uniprot_ids, cell_type_uniprot_ids, localization_uniprot_ids)\n",
      "\u001b[1;32mc:\\Users\\Wyss User\\Documents\\EVs\\OLINK\\10_analysis.ipynb Cell 11\u001b[0m line \u001b[0;36m2\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39midentify_targets\u001b[39m(correct_fractionation_uniprot_ids, cell_type_uniprot_ids, localization_uniprot_ids):\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/Wyss%20User/Documents/EVs/OLINK/10_analysis.ipynb#Y154sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m (\u001b[39mset\u001b[39m(correct_fractionation) \u001b[39m&\u001b[39m \u001b[39mset\u001b[39m(cell_type_uniprot_ids) \u001b[39m&\u001b[39m \u001b[39mset\u001b[39m(localization_uniprot_ids))\n",
      "\u001b[1;31mNameError\u001b[0m: name 'correct_fractionation' is not defined"
     ]
    }
   ],
   "source": [
    "def identify_fractionation_pattern_cell_type_specific_targets(\n",
    "    assay_dataframe, \n",
    "    uniprot_fasta_database, \n",
    "    region, \n",
    "    brain_rna_seq_dataframe, \n",
    "    cell_type, \n",
    "    general_specificity_metric, \n",
    "    general_specificity_cutoff, \n",
    "    indiv_specificity_metric, \n",
    "    indiv_specificity_cutoff,\n",
    "    tidy_dataframe,\n",
    "    high_fractions,\n",
    "    low_fractions,\n",
    "    sample_health,\n",
    "    mean_median_individual,\n",
    "    output_directory = 'olink_output', \n",
    "    ):\n",
    "    assays = pd.read_excel(ASSAY_LIST_PATH)\n",
    "\n",
    "    fasta_sequences = parse_gz_file(uniprot_fasta_database)\n",
    "    assays[\"Sequence\"] = assays[\"UniProt ID\"].map(lambda x: fasta_sequences.get(x, \"N/A\"))\n",
    "    localization_uniprot_ids = identify_localization(assay_dataframe, region)\n",
    "\n",
    "    cell_type_uniprot_ids = cell_type_enrichment(\n",
    "        brain_rna_seq_dataframe, cell_type, indiv_specificity_metric, indiv_specificity_cutoff, general_specificity_metric, general_specificity_cutoff\n",
    "    )\n",
    "\n",
    "    correct_fractionation_uniprot_ids = analyze_fractionation(tidy_dataframe, high_fractions, low_fractions, sample_health, mean_median_individual)\n",
    "\n",
    "    return identify_targets(correct_fractionation_uniprot_ids, cell_type_uniprot_ids, localization_uniprot_ids)\n",
    "    \n",
    "identify_fractionation_pattern_cell_type_specific_targets(assays, UNIPROT_TO_FASTA_PATH, \"internal\", brain_rna_seq, \"microglia\", \"tau\", 0.7, \"zscore\", 0.45, tidy_data, ['9', '10'], ['6', '11', '12', '13'], 'healthy', 'individual_median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "internal_dict = {\n",
    "    \"Microglia\": len(internal_microglia_uniprot_ids),\n",
    "    \"Astrocytes\": len(internal_astrocyte_uniprot_ids),\n",
    "    \"Oligodendrocytes\": len(internal_oligodendrocyte_uniprot_ids),\n",
    "    \"Neurons\": len(internal_neuron_uniprot_ids),\n",
    "}\n",
    "\n",
    "x_labels = list(internal_dict.keys())\n",
    "y_labels = list(internal_dict.values())\n",
    "\n",
    "plt.bar(x_labels, y_labels, color=\"black\")\n",
    "\n",
    "for i, v in enumerate(y_labels):\n",
    "    plt.text(i, v + 0.5, str(v), ha=\"center\", va=\"bottom\")\n",
    "\n",
    "plt.yticks([2, 4, 6, 8, 10, 12, 14])\n",
    "\n",
    "plt.xlabel(\"Cell Type\")\n",
    "plt.ylabel(\"Number of Targets\")\n",
    "plt.title(\"Internal Targets for Each Cell Type\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CSF_SAMPLES = [\n",
    "    \"SEC Fract 6 \",\n",
    "    \"SEC Fract 7\",\n",
    "    \"SEC Fract 8\",\n",
    "    \"SEC Fract 9\",\n",
    "    \"SEC Fract 10\",\n",
    "    \"SEC Fract 11\",\n",
    "    \"SEC Fract 12\",\n",
    "    \"SEC Fract 13\",\n",
    "    \"SEC Fract 14\",\n",
    "    \"SEC Fract 15\",\n",
    "    \"CSF A Internal EV\",\n",
    "    \"CSF B Internal EV\",\n",
    "    \"CSF C Internal EV\",\n",
    "    \"CSF D Internal EV\",\n",
    "]\n",
    "\n",
    "\n",
    "def graph_medians(uniprot_id):\n",
    "    df = tidy_data[uniprot_id]\n",
    "    df = df.reset_index(level=[\"SampleID\", \"Health\", \"Sample\"])\n",
    "\n",
    "    df[\"Sample\"] = pd.Categorical(df[\"Sample\"], categories=CSF_SAMPLES, ordered=True)\n",
    "\n",
    "    df_sorted = df.sort_values(\"Sample\")\n",
    "\n",
    "    grouped_data = [\n",
    "        group[uniprot_id].values for name, group in df_sorted.groupby(\"Sample\")\n",
    "    ]\n",
    "\n",
    "    plt.boxplot(grouped_data, notch=None, vert=None, patch_artist=None, widths=None)\n",
    "\n",
    "    plt.xlabel(\"Sample Description\")\n",
    "\n",
    "    plt.ylabel(\"Delta\")\n",
    "\n",
    "    plt.title(f\"{uniprot_id} Fractionation Pattern\")\n",
    "\n",
    "    plt.xticks(range(1, len(CSF_SAMPLES) + 1), CSF_SAMPLES)\n",
    "\n",
    "    plt.xticks(rotation=45, ha=\"right\")\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
